proposals/NNNN-compilation-flags-with-values.md

```diff
@@ -0,0 +1,63 @@
+
+## Motivation
+
+When building an application it is often useful to embed information in the binary to aid with support such as a build number, build date or version. It is also useful to build an application with particular tuneable parameters proved from the test/production build configuration such as a tolerance for results, sample rate of a signal or a URL. Up until now, this needed to be stored in the application's `Info.plist` but often it is not convenient to extract values from this resource and sometimes it is not desirable that parameter values be quite so public.
+
+## Proposed solution
+
+If this proposal is accepted, the existing Swift compiler frontend flag `"-D"` would be adapted so it is able to take values. Operation of the option without values would not be affected, the custom parameter being assumed to have a value `true` and if a value is provided it would be `true` from the point of view of conditionals unless it had the value `false`. Other values would be categorised as either integer, float or string values by a cursory parsing and be made available as a global symbol that can be referred to in expressions in a Swift source file across the module. They would not be actual Swift variables such as a fileprivate declaration but a form of "alias" as they are in other compilers where the expression will be parsed as if the user has entered that sequence of characters at that place in the source. This allows parameters to have an omni-type, determined late according to their context in an expression. For example, a compiler argument `-DNUMBER=10` could have `Int` or `Float` or `Double` type depending on it's context in the expression or declaration. 
+
+## Detailed design
+
+In terms of implementation this feature is implemented in two steps. When lexing an identifier, a check is made to see if it has the name of a custom compilation flag and if so, it's value extracted. If the extracted value is surrounded in `"` characters it is assumed to be a string literal or if the entirety of the parameter parses as an int the it is an int literal then likewise for float/double otherwise an error is emitted.
+
+If the type is recognised for a custom parameter, a literal token is created of that type but with the original source code range and a bit flag set on the token. During parsing, if a token is encountered which has been flagged as a custom compilation flag literal it is looked up and the value substituted as the "Text" of it's `AST` node as it is created but with the original source location. This was found to be the best way to have syntax highlighting (SourceKit) work as it should. If the token has string, int or float literal type, the reference to the flag is coloured appropriately in the Xcode source editor. For example a reference to a string parameter will be coloured red, a numeric flag blue rather than the green or black of a  normal identifier so the user has an indication it is an alias.
```

Comment:
- U.S. English spelling of "coloured"?
- Ditto "categorised" or "recognised"?

Reply:
Thanks for the review @benrimmington, I’ve pushed an amend commit. I don’t think I can generate a toolchain myself - can you do it for me?

Reply:
No, I haven't got commit access, sorry. One issue with swift-ci toolchains is they might expire. The review manager could build them when they start the review.